state/executor.go

```diff
@@ -301,80 +303,130 @@ func (e *BlockExecutor) execute(ctx context.Context, state types.State, block *t
 		if err != nil {
 			return nil, err
 		}
 		ISRs = append(ISRs, isr)
-		err = e.checkFraudProofTrigger(isr, currentIsrs, currentIsrIndex)
-		if err != nil {
-			return nil, err
+		isFraud := e.checkFraudProofTrigger(isr, currentIsrs, currentIsrIndex)
+		if isFraud {
+			fraudProof, err := e.generateFraudProof(&beginBlockRequest, nil, nil)
+			if err != nil {
+				return nil, err
+			}
+			// TODO: gossip fraudProof to P2P network
+			// fraudTx: BeginBlock
+			_ = fraudProof
 		}
 		currentIsrIndex++
 	}
+	deliverTxRequests := make([]*abci.RequestDeliverTx, len(block.Data.Txs))
+	for _, tx := range block.Data.Txs {
+		deliverTxRequest := abci.RequestDeliverTx{Tx: tx}
+		deliverTxRequests = append(deliverTxRequests, &deliverTxRequest)
+		res := e.proxyApp.DeliverTxAsync(deliverTxRequest)
+		if res.GetException() != nil {
+			return nil, errors.New(res.GetException().GetError())
+		}
 
-	abciResponses.EndBlock, err = e.proxyApp.EndBlockSync(abci.RequestEndBlock{Height: int64(block.Header.Height)})
-	if err != nil {
-		return nil, err
-	}
-	isr, err = e.getAppHash()
-	if err != nil {
-		return nil, err
+		if fraudProofsEnabled {
+			isr, err := e.getAppHash()
+			if err != nil {
+				return nil, err
+			}
+			ISRs = append(ISRs, isr)
+			isFraud := e.checkFraudProofTrigger(isr, currentIsrs, currentIsrIndex)
+			if isFraud {
+				fraudProof, err := e.generateFraudProof(&beginBlockRequest, deliverTxRequests, nil)
+				if err != nil {
+					return nil, err
+				}
+				// TODO: gossip fraudProof to P2P network
+				// fraudTx: current DeliverTx
+				_ = fraudProof
+			}
+			currentIsrIndex++
+		}
 	}
-	ISRs = append(ISRs, isr)
-	err = e.checkFraudProofTrigger(isr, currentIsrs, currentIsrIndex)
+	endBlockRequest := abci.RequestEndBlock{Height: int64(block.Header.Height)}
+	abciResponses.EndBlock, err = e.proxyApp.EndBlockSync(endBlockRequest)
 	if err != nil {
 		return nil, err
 	}
-	if block.Data.IntermediateStateRoots.RawRootsList == nil {
-		// Block producer: Initial ISRs generated here
-		block.Data.IntermediateStateRoots.RawRootsList = ISRs
+
+	if fraudProofsEnabled {
+		isr, err := e.getAppHash()
+		if err != nil {
+			return nil, err
+		}
+		ISRs = append(ISRs, isr)
+		isFraud := e.checkFraudProofTrigger(isr, currentIsrs, currentIsrIndex)
+		if isFraud {
+			fraudProof, err := e.generateFraudProof(&beginBlockRequest, deliverTxRequests, &endBlockRequest)
+			if err != nil {
+				return nil, err
+			}
+			// TODO: gossip fraudProof to P2P network
+			// fraudTx: EndBlock
+			_ = fraudProof
+		}
+		if block.Data.IntermediateStateRoots.RawRootsList == nil {
+			// Block producer: Initial ISRs generated here
+			block.Data.IntermediateStateRoots.RawRootsList = ISRs
+		}
 	}
 
 	return abciResponses, nil
 }
 
-func (e *BlockExecutor) checkFraudProofTrigger(generatedIsr []byte, currentIsrs [][]byte, index int) error {
+func (e *BlockExecutor) checkFraudProofTrigger(generatedIsr []byte, currentIsrs [][]byte, index int) bool {
```

Comment:
Instead of passing `[][]byte` and `index`, you can pass element of `currentIsrs` directly.